File: RainyClock/Services/AlarmTimeCalculator.swift
import Foundation

enum AlarmTimeCalculator {
    /// Moves a `Calendar` weekday (1 = Sunday ... 7 = Saturday) by whole days,
    /// wrapping around the week in either direction.
    ///
    /// Both schedulers and the scheduled-alarm summary need this: a rain lead time
    /// that crosses midnight puts the ring on the previous day, so every selected
    /// weekday shifts with it.
    static func shiftedWeekday(_ weekday: Int, byDays dayShift: Int) -> Int {
        ((weekday - 1 + dayShift) % 7 + 7) % 7 + 1
    }

    static func nextAlarmDateForWeatherCheck(
        alarmTime: Date,
        leadTimeMinutes: Int,
        shouldApplyLeadTime: Bool,
        rainProbabilityThreshold: Double,
        maximumPrecipitationProbability: Double,
        selectedWeekdays: Set<Int> = CommuteAlarmSettings.allWeekdays,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> ScheduledAlarmSummary {
        let timeComponents = calendar.dateComponents([.hour, .minute], from: alarmTime)
        let selectedWeekdays = selectedWeekdays.isEmpty ? CommuteAlarmSettings.allWeekdays : selectedWeekdays
        let normalAlarmDate = nextDate(
            hour: timeComponents.hour ?? 7,
            minute: timeComponents.minute ?? 30,
            selectedWeekdays: selectedWeekdays,
            after: now,
            calendar: calendar
        ) { candidate in
            let weatherRefreshDate = calendar.date(byAdding: .minute, value: -leadTimeMinutes, to: candidate) ?? candidate
            return weatherRefreshDate > now
        }

        let weatherRefreshDate = calendar.date(byAdding: .minute, value: -leadTimeMinutes, to: normalAlarmDate) ?? normalAlarmDate
        let scheduledAlarmDate = shouldApplyLeadTime
            ? weatherRefreshDate
            : normalAlarmDate

        return ScheduledAlarmSummary(
            normalAlarmDate: normalAlarmDate,
            scheduledAlarmDate: scheduledAlarmDate,
            weatherRefreshDate: weatherRefreshDate,
            exceedsRainThreshold: shouldApplyLeadTime,
            leadTimeMinutes: shouldApplyLeadTime ? leadTimeMinutes : 0,
            rainProbabilityThreshold: rainProbabilityThreshold,
            maximumPrecipitationProbability: maximumPrecipitationProbability
        )
    }

    private static func nextDate(
        hour: Int,
        minute: Int,
        selectedWeekdays: Set<Int>,
        after now: Date,
        calendar: Calendar
    ) -> Date {
        nextDate(
            hour: hour,
            minute: minute,
            selectedWeekdays: selectedWeekdays,
            after: now,
            calendar: calendar
        ) { candidate in
            candidate > now
        }
    }

    private static func nextDate(
        hour: Int,
        minute: Int,
        selectedWeekdays: Set<Int>,
        after now: Date,
        calendar: Calendar,
        isValidCandidate: (Date) -> Bool
    ) -> Date {
        let todayStart = calendar.startOfDay(for: now)

        for dayOffset in 0...7 {
            guard let day = calendar.date(byAdding: .day, value: dayOffset, to: todayStart),
                  let candidate = calendar.date(bySettingHour: hour, minute: minute, second: 0, of: day) else {
                continue
            }

            guard isValidCandidate(candidate) else {
                continue
            }

            if selectedWeekdays.contains(calendar.component(.weekday, from: candidate)) {
                return candidate
            }
        }

        return now
    }
}

File: android/app/src/main/java/com/shukaihu/rainyclock/logic/AlarmTimeCalculator.kt
package com.shukaihu.rainyclock.logic

import com.shukaihu.rainyclock.model.CommuteAlarmSettings
import com.shukaihu.rainyclock.model.ScheduledAlarmSummary
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

/**
 * Port of the iOS `AlarmTimeCalculator`. Weekdays use the iOS `Calendar`
 * numbering (1 = Sunday … 7 = Saturday); [sundayBasedWeekday] converts from
 * java.time's ISO numbering.
 */
object AlarmTimeCalculator {

    fun sundayBasedWeekday(dateTime: ZonedDateTime): Int =
        dateTime.dayOfWeek.value % 7 + 1

    /**
     * Moves a weekday (1 = Sunday … 7 = Saturday) by whole days, wrapping
     * around the week in either direction.
     */
    fun shiftedWeekday(weekday: Int, dayShift: Int): Int =
        ((weekday - 1 + dayShift) % 7 + 7) % 7 + 1

    /**
     * Picks the next alarm occurrence whose weather-refresh point is still in
     * the future, then applies the rain decision to it. Mirrors
     * `nextAlarmDateForWeatherCheck` on iOS.
     */
    fun nextAlarmDateForWeatherCheck(
        alarmHour: Int,
        alarmMinute: Int,
        leadTimeMinutes: Int,
        shouldApplyLeadTime: Boolean,
        rainProbabilityThreshold: Double,
        maximumPrecipitationProbability: Double,
        selectedWeekdays: Set<Int> = CommuteAlarmSettings.ALL_WEEKDAYS,
        now: Instant = Instant.now(),
        zone: ZoneId = ZoneId.systemDefault()
    ): ScheduledAlarmSummary {
        val weekdays = selectedWeekdays.ifEmpty { CommuteAlarmSettings.ALL_WEEKDAYS }
        val normalAlarmDate = nextDate(alarmHour, alarmMinute, weekdays, now, zone) { candidate ->
            candidate.minusMinutes(leadTimeMinutes.toLong()).toInstant().isAfter(now)
        }

        val weatherRefreshDate = normalAlarmDate.minusSeconds(leadTimeMinutes * 60L)
        val scheduledAlarmDate = if (shouldApplyLeadTime) weatherRefreshDate else normalAlarmDate

        return ScheduledAlarmSummary.of(
            normalAlarmDate = normalAlarmDate,
            scheduledAlarmDate = scheduledAlarmDate,
            weatherRefreshDate = weatherRefreshDate,
            exceedsRainThreshold = shouldApplyLeadTime,
            leadTimeMinutes = if (shouldApplyLeadTime) leadTimeMinutes else 0,
            rainProbabilityThreshold = rainProbabilityThreshold,
            maximumPrecipitationProbability = maximumPrecipitationProbability
        )
    }

    private fun nextDate(
        hour: Int,
        minute: Int,
        selectedWeekdays: Set<Int>,
        now: Instant,
        zone: ZoneId,
        isValidCandidate: (ZonedDateTime) -> Boolean
    ): Instant {
        val todayStart = now.atZone(zone).toLocalDate()

        for (dayOffset in 0..7) {
            val day = todayStart.plusDays(dayOffset.toLong())
            val candidate = day.atTime(hour, minute).atZone(zone)
            if (!isValidCandidate(candidate)) continue
            if (selectedWeekdays.contains(sundayBasedWeekday(candidate))) {
                return candidate.toInstant()
            }
        }

        return now
    }
}
